stepper-lib/src/air/call/utils.rs

```diff
@@ -40,9 +40,34 @@ pub(super) fn set_local_call_result<'i>(
     match output {
         CallOutput::Scalar(name) => {
             match exec_ctx.data_cache.entry(name.to_string()) {
-                Vacant(entry) => entry.insert(AValue::JValueRef(result)),
-                Occupied(entry) => return Err(MultipleVariablesFound(entry.key().clone())),
+                Vacant(entry) => {
+                    entry.insert(AValue::JValueRef(result));
+                }
+                Occupied(mut entry) => {
+                    // check that current execution flow is inside a fold block
+                    if exec_ctx.met_folds.is_empty() {
+                        // shadowing is allowed only inside fold blocks
+                        return Err(MultipleVariablesFound(entry.key().clone()));
+                    }
+
+                    match entry.get() {
+                        AValue::JValueRef(_) => {}
+                        // shadowing is allowed only for scalar values
+                        _ => return Err(MultipleVariablesFound(entry.key().clone())),
+                    };
+
+                    entry.insert(AValue::JValueRef(result));
+                }
             };
+
+            if let Some(fold_block_name) = exec_ctx.met_folds.back() {
+                let fold_state = match exec_ctx.data_cache.get_mut(*fold_block_name) {
+                    Some(AValue::JValueFoldCursor(fold_state)) => fold_state,
+                    _ => unreachable!(),
```

Comment:
Maybe add a message, in case it's reached?

```suggestion
                    _ => unreachable!("fold block data must be represented as fold cursor"),
```